### PR TITLE
19760 - initial partial refunds model persistence

### PR DESCRIPTION
**Pull request by ochiu**
*Issue #:*
https://github.com/bcgov/entity/issues/19760
*Description of changes:*
- update refund schema - add optional refundRevenue
- persist refund partial records
- extend converter to support camel to snake case conversion of properties
- extend converter to handle enum values vs enum objects

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

File: pay-api/src/pay_api/services/refund.py [hunk 1]
```diff
@@ -268,12 +271,43 @@ def create_refund(cls, invoice_id: int, request: Dict[str, str], **kwargs) -> Di
...
+            db.session.add(refund_line)
+
+    @staticmethod
+    def _get_partial_refund_lines(refund_revenue: List[Dict]) -> List[RefundPartialLine]:
+        """Convert Refund revenue data to a list of Partial Refund lines."""
+        if not refund_revenue:
+            return []
+
+        refund_lines = []
+        for refund in refund_revenue:
```
Comment (ochiu) [inline pay-api/src/pay_api/services/refund.py:305]: Investigating an issue when using the converter where the payload is camel case but the data class is snake case.

Trying to update the converter to allow for a parameter to do a camel to snake case conversion when it is instantiated.

File: pay-api/src/pay_api/utils/converter.py [hunk 2]
```diff
@@ -1,25 +1,76 @@
...
 import cattrs
+from attrs import fields, has
+from cattrs.gen import make_dict_structure_fn, make_dict_unstructure_fn, override
 
 
 class Converter(cattrs.Converter):
     """Addon to cattr converter."""
 
-    def __init__(self):
+    def __init__(self, camel_to_snake_case: bool = False, enum_to_value: bool = False):
```
Comment (ochiu) [inline pay-api/src/pay_api/utils/converter.py:15]: We should probably convert this into a data class for ConverterOptions if we start adding more configurations in the future

File: pay-api/src/pay_api/services/refund.py (hunk 1 above)
Comment (ochiu) [reply pay-api/src/pay_api/services/refund.py]: Updated converter to handle camel to snake conversion

**Review by seeker25 — APPROVED** [review]

File: pay-api/src/pay_api/models/refunds_partial.py [hunk 3]
```diff
@@ -50,6 +52,19 @@ class RefundsPartial(Audit, VersionedModel):  # pylint: disable=too-many-instanc
     refund_amount = db.Column(db.Numeric(19, 2), nullable=False)
     refund_type = db.Column(db.String(50), nullable=True)
 
+    @classmethod
+    def find_by_invoice_id(cls, invoice_id: int):
+        """Return refund partials by invoice id."""
+        return db.session.query(RefundsPartial)\
+            .join(PaymentLineItem, PaymentLineItem.id == RefundsPartial.payment_line_item_id)\
+            .filter(PaymentLineItem.invoice_id == invoice_id).all()
```
Comment (seeker25) [inline pay-api/src/pay_api/models/refunds_partial.py:60]: Probably better to put this in the service and keep the models barebones?

You reference to payment_line_item in the model, which could cause some pain of circular dependencies in the future

Comment (seeker25) [reply pay-api/src/pay_api/models/refunds_partial.py:60]: I wonder if there is a way to reference to PaymentLineItem, without having to include it at the top somehow

Comment (ochiu) [reply pay-api/src/pay_api/models/refunds_partial.py]: agreed, good call

**Pull request merged**